File: patterns/python/fast_and_slow_pointers.py
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class FastAndSlowPointers:
    # LeetCode 141 - Linked List Cycle (HashSet Approach)
    def hasCycleHashSetApproach(self, head):
        visited = set()
        current = head
        while current:
            if current in visited:
                return True
            visited.add(current)
            current = current.next
        return False

    # LeetCode 141 - Linked List Cycle (Fast and Slow Pointer Approach)
    def hasCycleFastAndSlowPointersApproach(self, head):
        if not head or not head.next:
            return False
        slow, fast = head, head
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
            if slow == fast:
                return True
        return False

    # LeetCode 876 - Middle of the Linked List (Counting Approach)
    def middleNodeCountingApproach(self, head):
        count = 0
        current = head
        while current:
            count += 1
            current = current.next
        current = head
        for _ in range(count // 2):
            current = current.next
        return current

    # LeetCode 876 - Middle of the Linked List (Fast and Slow Pointer Approach)
    def middleNodeFastAndSlowPointerApproach(self, head):
        slow, fast = head, head
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
        return slow

    # LeetCode 202 - Happy Number (HashSet Approach)
    def getSumOfSquares(self, n):
        return sum(int(digit)**2 for digit in str(n))

    def isHappyHashSetApproach(self, n):
        seen = set()
        while n != 1 and n not in seen:
            seen.add(n)
            n = self.getSumOfSquares(n)
        return n == 1

    # LeetCode 202 - Happy Number (Fast and Slow Pointer Approach)
    def isHappyFastAndSlowPointersApproach(self, n):
        slow = n
        fast = self.getSumOfSquares(n)
        while fast != 1 and slow != fast:
            slow = self.getSumOfSquares(slow)
            fast = self.getSumOfSquares(self.getSumOfSquares(fast))
        return fast == 1

File: patterns/c++/FastAndSlowPointers.cpp
#include <unordered_set>
using namespace std;

class ListNode {
public:
    int val;
    ListNode* next;
    ListNode(int x) : val(x), next(nullptr) {}
};

class FastAndSlowPointers {
public:
    // LeetCode 141 - Linked List Cycle (HashSet Approach)
    bool hasCycleHashSetApproach(ListNode* head) {
        unordered_set<ListNode*> visited;
        ListNode* current = head;
        while (current != nullptr) {
            if (visited.find(current) != visited.end()) {
                return true; // Cycle detected
            }
            visited.insert(current);
            current = current->next;
        }
        return false; // No cycle
    }

    // LeetCode 141 - Linked List Cycle (Fast and Slow Pointer Approach)
    bool hasCycleFastAndSlowPointersApproach(ListNode* head) {
        if (head == nullptr || head->next == nullptr) {
            return false;
        }
        ListNode* slow = head;
        ListNode* fast = head;
        while (fast != nullptr && fast->next != nullptr) {
            slow = slow->next;
            fast = fast->next->next;
            if (slow == fast) {
                return true; // Cycle detected
            }
        }
        return false; // No cycle
    }

    // LeetCode 876 - Middle of the Linked List (Counting Approach)
    ListNode* middleNodeCountingApproach(ListNode* head) {
        int count = 0;
        ListNode* current = head;
        while (current != nullptr) {
            count++;
            current = current->next;
        }
        current = head;
        for (int i = 0; i < count / 2; i++) {
            current = current->next;
        }
        return current;
    }

    // LeetCode 876 - Middle of the Linked List (Fast and Slow Pointer Approach)
    ListNode* middleNodeFastAndSlowPointerApproach(ListNode* head) {
        ListNode* slow = head;
        ListNode* fast = head;
        while (fast != nullptr && fast->next != nullptr) {
            slow = slow->next;
            fast = fast->next->next;
        }
        return slow;
    }

    // LeetCode 202 - Happy Number (HashSet Approach)
    int getSumOfSquares(int n) {
        int sum = 0;
        while (n > 0) {
            int digit = n % 10;
            sum += digit * digit;
            n /= 10;
        }
        return sum;
    }

    bool isHappyHashSetApproach(int n) {
        unordered_set<int> seen;
        while (n != 1 && seen.find(n) == seen.end()) {
            seen.insert(n);
            n = getSumOfSquares(n);
        }
        return n == 1;
    }

    // LeetCode 202 - Happy Number (Fast and Slow Pointer Approach)
    bool isHappyFastAndSlowPointersApproach(int n) {
        int slow = n;
        int fast = getSumOfSquares(n);
        while (fast != 1 && slow != fast) {
            slow = getSumOfSquares(slow);
            fast = getSumOfSquares(getSumOfSquares(fast));
        }
        return fast == 1;
    }
};